jujugui/static/gui/src/app/state/state.js

```diff
@@ -82,6 +95,76 @@ const State = class State {
   }
 
   /**
+    Check the supplied block of path to see if this is a user store path or
+    not. These options are becuase the store path can come in many forms and
+    we need to know if it's a user store path for both parsing and
+    generating the path.
+    @return {Boolean}
+  */
+  _isUserStorePath(block) {
+    const seriesList = this.seriesList;
+    // Store path ex) django/bundle/42
+    if (block.length === 3 && block[1] === 'bundle') {
+      return false;
+    }
+    // Store path ex) hatch/ghost
+    if (block.length === 2 && !seriesList.includes(block[1])) {
+      return Number.isNaN(parseInt(block[1], 10));
+    }
+    // Everything else
+    return !Number.isNaN(parseInt(block[2], 10)) ||
+      seriesList.includes(block[2]);
+  }
+
+  /**
+    The current window location. If set has been called on this property then
+    it will return the externally set option. Typically this option will be used
+    for testing.
+    @type {Object}
+  */
+  get location() {
+    return this._location || window.location;
+  }
+
+  set location(location) {
+    this._location = location;
+  }
+
+  /**
+    The object representing the current app state.
+    @type {Object}
+  */
+  get appState() {
+    return this._appStateHistory[this._appStateHistory.length-1];
+  }
+
+  /**
+    Grabs the current location and parses it, building the state, then
+    executing the registered dispatchers.
+  */
+  dispatch() {
+    let error, state;
+    ({error, state} = this.generateState(this.location.href));
+    this._appStateHistory.push(state);
+    if (error !== null) {
```

Comment:
I'd just assume that, in case of errors, no usable state is ever returned. Otherwise it could be confusin, instead "it's either a good state or a good error" is more simple to explain. What do you think?